File: examples/cp2ds_cpp/native/coreprofiles2distsource.cpp
#include "UALClasses.h"


void coreprofiles2distsource_cpp(IdsNs::IDS::core_profiles& in_core_profiles, IdsNs::IDS::distribution_sources& out_distribution_sources, int* status_code, char** status_message)
{
    const char* TEXT = "Status info of coreprofiles2distsource CPP";
    int idsSize = -1;

    printf("Entering subroutine eq2dist");
    idsSize = in_core_profiles.time.extent(0);

    printf("size of input IDS  = .d", idsSize);

    // INITIALISATION OF STATUS INFO
    *status_message = (char*)malloc(strlen(TEXT) + 1);
    strcpy (*status_message, TEXT);

    out_distribution_sources.time.resize(idsSize);

    // Fill in the output IDS (Physical data)
    for(int i=0; i < idsSize; i++)
    {
        // Time : copy from input IDS
        printf("Received input time from equilibrium : %f", in_core_profiles.time(i));
        out_distribution_sources.time(i) = in_core_profiles.time(i);
    }

    out_distribution_sources.ids_properties.homogeneous_time = 1;
    out_distribution_sources.code.name   = "coreprofiles2distsource_cpp";
    out_distribution_sources.code.version   = "1.0";
    out_distribution_sources.code.parameters = "my_code_specific_parameters";
    out_distribution_sources.code.output_flag = 0   ;

    *status_code = 0;
    printf("End of coreprofiles2distsource CPP");

}

File: examples/cp2ds/native/coreprofiles2distsource.f90
module mod_coreprofiles2distsource
contains
subroutine coreprofiles2distsource(coreprofilesin, distsourceout, error_flag, error_message)

use ids_schemas

implicit none

integer,parameter :: DP=kind(1.0D0)

type (ids_core_profiles) :: coreprofilesin
type (ids_distribution_sources) :: distsourceout
integer, intent(out) :: error_flag
character(len=:), pointer, intent(out) :: error_message

integer :: i

write(0,*) 'Entering subroutine eq2dist'

write(0,*) 'size of input IDS  = ',size(coreprofilesin%time)

  ! INITIALISATION OF ERROR FLAG
  error_flag = 0
  allocate(character(50):: error_message)
  error_message = 'Status info of coreprofiles2distsource'

! The output IDS  must be allocated with its number of time slices (1 for a single time slice physics module)
! Here we allocate the output IDS  to the same size as the input IDS (but this is not a general rule)
allocate(distsourceout%time(size(coreprofilesin%time)))

! Fill in the output IDS (Physical data)
do i=1,size(coreprofilesin%time)
   ! Time : copy from input IDS
   write(0,*) 'Received input time from equilibrium : ', coreprofilesin%time(i)
   distsourceout%time(i) = coreprofilesin%time(i)  
   ! THE TIME FIELD MUST BE FILLED (MANDATORY) in case of multiple time slice mode for the IDS;
    
enddo

distsourceout%ids_properties%homogeneous_time = 1

allocate(distsourceout%code%name(1))   ! For a string of 132 characters max.
distsourceout%code%name(1)   = 'equ2dist'
allocate(distsourceout%code%version(1))   ! For a string of 132 characters max.
distsourceout%code%version(1)   = '1.0'
allocate(distsourceout%code%parameters(1))   ! For a string of 132 characters max.
distsourceout%code%parameters(1) = 'my_code_specific_parameters'
distsourceout%code%output_flag = 0   ! Integer output flag, 0 means the run was successful and can be used in the rest of the workflow, <0 means failure

write(0,*) 'End Subroutine'

return
end subroutine
end module mod_coreprofiles2distsource
